rust/src/sign/mod.rs

```diff
@@ -0,0 +1,325 @@
+    pub fn body(self, body: &[u8]) -> SignerBuilder<'a, K, Pk, &[u8], Method, Path> {
+        SignerBuilder {
+            kid: self.kid,
+            private_key: self.private_key,
+            body,
+            method: self.method,
+            path: self.path,
+            headers: self.headers,
+            jws_jku: self.jws_jku,
+        }
+    }
+}
+
+impl<'a, K, Pk, Body, Path> SignerBuilder<'a, K, Pk, Body, Unset, Path> {
+    /// Add the request method.
+    pub fn method<M>(self) -> SignerBuilder<'a, K, Pk, Body, PhantomData<M>, Path> {
+        SignerBuilder {
+            kid: self.kid,
+            private_key: self.private_key,
+            body: self.body,
+            method: PhantomData,
+            path: self.path,
+            headers: self.headers,
+            jws_jku: self.jws_jku,
+        }
+    }
+}
+
+impl<'a, K, Pk, Body, Method> SignerBuilder<'a, K, Pk, Body, Method, Unset> {
+    /// Add the request absolute path starting with a leading `/` and without
+    /// any trailing slashes.
+    pub fn path(self, path: &str) -> SignerBuilder<'a, K, Pk, Body, Method, &str> {
+        assert!(
+            path.starts_with('/'),
+            "Invalid path \"{path}\" must start with '/'"
+        );
+        SignerBuilder {
+            kid: self.kid,
+            private_key: self.private_key,
+            body: self.body,
+            method: self.method,
+            path,
+            headers: self.headers,
+            jws_jku: self.jws_jku,
+        }
+    }
+}
+
+impl<'a, K, Pk, Body, Method, Path> SignerBuilder<'a, K, Pk, Body, Method, Path> {
+    /// Add a header name & value.
+    /// May be called multiple times to add multiple different headers.
+    ///
+    /// Warning: Only a single value per header name is supported.
+    pub fn header(mut self, key: &'a str, value: &'a [u8]) -> Self {
+        self.headers.insert(HeaderName(key), value);
+        self
+    }
+
+    /// Appends multiple header names & values.
+    ///
+    /// Warning: Only a single value per header name is supported.
+    pub fn headers(mut self, headers: impl IntoIterator<Item = (&'a str, &'a [u8])>) -> Self {
+        self.headers
+            .extend(headers.into_iter().map(|(k, v)| (HeaderName(k), v)));
+        self
+    }
+
+    /// Sets the jws header `jku` JSON Web Key URL.
+    ///
+    /// Note: This is not generally required when calling APIs,
+    /// but is set on webhook signatures.
+    pub fn jku(mut self, jku: &'a str) -> Self {
+        self.jws_jku = Some(jku);
+        self
+    }
+}
+
+impl<'a> SignerBuilder<'a, &'a str, Unset, Unset, PhantomData<Get>, &'a str> {
+    /// Builds a [`CustomSigner`]
+    pub fn build_custom_signer(self) -> CustomSigner<'a> {
+        CustomSigner {
+            kid: self.kid,
+            body: &[],
+            method: Get::name(),
+            path: self.path,
+            headers: self.headers,
+            jws_jku: self.jws_jku,
+        }
+    }
+}
```

Comment:
If we're getting rid of the old `sign_with_pem` and `verify_with_...` functions and going for the full breaking change, do we still need the distinction between old (v1) and new "custom" signers/verifiers?

Just wondering if they can be combined and just expose different functions (depending on which types are populated in the builder).